com/williamfiset/datastructures/binarysearchtree/SplayTree.java

```diff
@@ -0,0 +1,274 @@
+            } else if (root.getRight().getData().compareTo(key) < 0)// Zag-Zag (Right Right)
+            {
+                root.getRight().setRight(splayUtil(root.getRight().getRight(), key));
+                root = leftRotate(root);
+            }
+
+            return (root.getRight() == null) ? root : leftRotate(root);
+        }
+
+    }
+
+    public BinaryTree<T> splay(T x) {
+        if (root == null) return null;
+
+        this.root = splayUtil(root, x);
+
+        return this.root;
+    }
+    public BinaryTree<T> search(T x) {
+        if (root == null) return null;
+
+        this.root = splayUtil(root, x);
+
+        return this.root.getData().compareTo(x)==0 ? this.root:null;
+    }
+
+    private ArrayList<BinaryTree<T>> split(T x) {
+        BinaryTree<T> right;
+        BinaryTree<T> left;
+
+        if (x.compareTo(root.getData()) > 0) {
+            right = root.getRight();
+            left = root;
+            left.setRight(null);
+        } else {
+            left = root.getLeft();
+            right = root;
+            right.setLeft(null);
+        }
+        ArrayList<BinaryTree<T>> l_r = new ArrayList<>();
+        l_r.add(left);
+        l_r.add(right);
+
+        return l_r;
+    }
+
+    public BinaryTree<T> insert(T x) {
+        if (root == null) {
+            root = new BinaryTree<>(x);
+            return root;
+        }
+        splay(x);
+
+        ArrayList<BinaryTree<T>> l_r = split(x);
+
+        BinaryTree<T> left = l_r.get(0);
+        BinaryTree<T> right = l_r.get(1);
+
+        root = new BinaryTree<>(x);
+        root.setLeft(left);
+        root.setRight(right);
+
+
+        return root;
+    }
+
+    public BinaryTree<T> delete(T x) {
+        if (root == null) return null;
+
+        BinaryTree<T> searchResult = splay(x);
+
+        if (searchResult.getData().compareTo(x) != 0)
+            return null;
+
+        BinaryTree<T> leftSubtree = root.getLeft();
+        BinaryTree<T> rightSubtree = root.getRight();
+
+        // Set the 'to be deleted' key ready for garbage collection
+        root.setLeft(null);
+        root.setRight(null);
+        root.setData(null);
+
+        root = join(leftSubtree, rightSubtree);
+
+        return root;
+    }
+
+    private BinaryTree<T> join(BinaryTree<T> L, BinaryTree<T> R) {
+
+        if (L == null) {
+            root = R;
+            return R;
+        }
+        root = splayUtil(L, findMax(L));
+        root.setRight(R);
+        return root;
+    }
+    public T findMax(BinaryTree<T> root){
+        BinaryTree<T> temp = root;
+        while (temp.getRight()!=null)
+            temp=temp.getRight();
+        return temp.getData();
+    }
+
+    @Override
+    public String toString() {
+
+        System.out.println("Elements:"+inorder(root, new ArrayList<>()));
+        return (root != null) ? root.toString() : null;
+    }
+
+    public ArrayList<T> inorder(BinaryTree<T> root, ArrayList<T> sorted) {
```

Comment:
Can you document how this is supposed to work and why someone might want to call inorder?

Reply:
yes, i should've made it private. Its not mandatory as well. It just shows all the elements before displaying the tree.